File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2015/Y15D04.kt
package org.gristle.pdxpuzzles.advent.y2015

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.math.md5

class Y15D04(private val input: String) : Day {
    private fun solve(digitLength: Int) = generateSequence(1) { it + 1 }
        .first { i -> (input + i).md5().take(digitLength).all { it == '0' } }

    override fun part1() = solve(5)
    override fun part2() = solve(6)
}

fun main() = Day.runDay(Y15D04::class)

//    Class creation: 16ms
//    Part 1: 117946 (103ms)
//    Part 2: 3938038 (1343ms)
//    Total time: 1463ms

File: rust/advent/src/bin/2015/y15d04.rs
use advent::utilities::get_input::get_input;
use md5::{Digest, Md5};
use utilities::structs::stopwatch::{ReportDuration, Stopwatch};

type Input<'a> = &'a str;
type Output = usize;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(15, 4).unwrap();
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn solve(salt: Input, digit_length: usize) -> Output {
    let mut hasher = Md5::new();
    let mut buf = [0u8; 32];
    (1..)
        .find(|i| {
            let message = format!("{}{}", salt, i);
            hasher.update(&message);
            base16ct::lower::encode_str(&hasher.finalize_reset(), &mut buf)
                .unwrap()
                .as_bytes()
                .iter()
                .take(digit_length)
                .all(|b| *b == b'0')
        })
        .unwrap()

}
fn part1(salt: Input) -> Output {
    solve(salt, 5)
}

fn part2(salt: Input) -> Output {
    solve(salt, 6)
}

#[test]
fn default() {
    let input = get_input(15, 4).unwrap();
    assert_eq!(117946, part1(&input));
    assert_eq!(3938038, part2(&input));
}

    // Input parsed (16μs)
    // 1. 117946 (17ms)
    // 2. 3938038 (542ms)
    // Total: 559ms